mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1167,6 +1197,223 @@ pub async fn p2sh_spending_tx<T: UtxoCommonOps>(coin: &T, input: P2SHSpendingTxI
     })
 }
 
+pub type GenDexFeeSpendResult = MmResult<TransactionInputSigner, TxGenError>;
+
+enum CalcPremiumBy {
+    DeductMinerFee,
+    UseExactAmount(u64),
+}
+
+async fn gen_dex_fee_spend_preimage<T: UtxoCommonOps>(
+    coin: &T,
+    args: &GenDexFeeSpendArgs<'_>,
+    lock_time: LocktimeSetting,
+    calc_premium: CalcPremiumBy,
+) -> GenDexFeeSpendResult {
+    let mut prev_tx: UtxoTx =
+        deserialize(args.dex_fee_tx).map_to_mm(|e| TxGenError::TxDeserialization(e.to_string()))?;
+    prev_tx.tx_hash_algo = coin.as_ref().tx_hash_algo;
+    drop_mutability!(prev_tx);
+
+    let dex_fee_sat = sat_from_big_decimal(&args.dex_fee_amount, coin.as_ref().decimals)?;
+    let premium_sat = match calc_premium {
+        CalcPremiumBy::UseExactAmount(sat) => sat,
+        CalcPremiumBy::DeductMinerFee => {
+            let miner_fee = coin
+                .get_htlc_spend_fee(DEFAULT_SWAP_TX_SPEND_SIZE, &FeeApproxStage::WithoutApprox)
+                .await?;
+
+            let premium_sat = sat_from_big_decimal(&args.premium_amount, coin.as_ref().decimals)?;
+            if miner_fee + coin.as_ref().dust_amount > premium_sat {
+                return MmError::err(TxGenError::MinerFeeExceedsPremium {
+                    miner_fee: big_decimal_from_sat_unsigned(miner_fee, coin.as_ref().decimals),
+                    premium: args.premium_amount.clone(),
+                });
+            }
+            premium_sat - miner_fee
+        },
+    };
+
+    let dex_fee_address = address_from_raw_pubkey(
+        args.dex_fee_pub,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.checksum_type,
+        coin.as_ref().conf.bech32_hrp.clone(),
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(|e| TxGenError::AddressDerivation(format!("Failed to derive dex_fee_address: {}", e)))?;
+    let dex_fee_output = TransactionOutput {
+        value: dex_fee_sat,
+        script_pubkey: Builder::build_p2pkh(&dex_fee_address.hash).to_bytes(),
+    };
+
+    let premium_address = address_from_raw_pubkey(
+        args.maker_pub,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.checksum_type,
+        coin.as_ref().conf.bech32_hrp.clone(),
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(|e| TxGenError::AddressDerivation(format!("Failed to derive premium_address: {}", e)))?;
+    let premium_output = TransactionOutput {
+        value: premium_sat,
+        script_pubkey: Builder::build_p2pkh(&premium_address.hash).to_bytes(),
```

Comment:
It seems that segwit/p2wpkh is not supported yet. Just reminding you @artemii235 for next PRs. I understand that there is no way to know if the other side is using segwit coin or not at the swaps layer, only information about that is in the orderbook so this will require some work.

Reply:
Yeah, it requires additional work, I will do it in next PRs :)

Reply:
@shamardy As `SIGHASH_SINGLE` is now implemented for taker signature, maker's output is now fully generated on maker's side, so this seems to be also resolved 🙂